Update DOCKER.md to include user-friendly steps of running service in Docker (#617)
elastic/connectors-python: docs/DOCKER.md

Download can be done manually, or just using the line below - make sure to update `--output` argument value if your directory name is different or you want to use a different config file name.

```sh
curl https://raw.githubusercontent.com/elastic/connectors-python/main/config.yml --output ~/connectors-python-config/config.yml
```

3. Update the configuration downloaded to make it work for your setup.

In case you're running Connector Service against a dockerised version of Elasticsearch and Kibana, your config file might look the following:

```
elasticsearch:
  host: http://host.docker.internal:9200
  username: elastic
  password: <YOUR_PASSWORD>
  ssl: true
  bulk:
    queue_max_size: 1024
    queue_max_mem_size: 25
    display_every: 100
    chunk_size: 1000
    max_concurrency: 5
    chunk_max_mem_size: 5
    concurrent_downloads: 10
  request_timeout: 120
  max_wait_duration: 120
  initial_backoff_duration: 1
  backoff_multiplier: 2
  log_level: info

service:
  idling: 30
  heartbeat: 300
  max_errors: 20
  max_errors_span: 600
  max_concurrent_syncs: 1
  job_cleanup_interval: 300
  log_level: INFO

> native_service_types:

Reviewer: Running the connector service in native mode should not be the default for on-prem - I remember we had a GH issue (not finding it atm though) to let users focus on running connector clients via _Build a connector_. 

Could we provide this example with `service_type` so that the default is to run connector clients via _Build a connector_?

Reviewer: Correct, let's focus on _Build a connector_ for on-prem. A short note about running the connector in native mode should be enough.

Author: I've updated the file, can you take a look @danajuratoni?

I've included a whole config for native connectors, though, should I drop it as well?

Author: This section of config file is now for native connectors (for on prem, look at the section above).

Do you mean I leave the whole config for native connectors out, just mentioning that the config downloaded from our repo is native by default?

Reviewer 2: The connector client config is commented out:
https://github.com/elastic/connectors-python/blob/0718287634ccdcc0da8407535b84aa3598570d6f/config.yml#L44-L46

I'm suggesting we also comment out `native_service_types`:
https://github.com/elastic/connectors-python/blob/0718287634ccdcc0da8407535b84aa3598570d6f/config.yml#L29-L41

For both native mode and connector client mode, you don't have to give the example of the whole config file in the MD.

E.g. 

```markdown
3. Update the configuration downloaded to make it work for your [on-prem connector](https://www.elastic.co/guide/en/enterprise-search/current/build-connector.html#build-connector-usage)

Uncomment the `connector_id` and `service_type`:

# connector information
connector_id: <CONNECTOR_ID_FROM_KIBANA>
service_type: '<DESIRED_SERVICE_TYPE>'
```

```markdown
In case you want to run connector service in native mode, you will need slightly different configuration file, having `native_service_types` config option populated instead of `connector_id` and `service_type`:

# remove entries from this list to not run these connectors in this instance of the service
native_service_types:
  - mongodb
  - mysql
```

Reviewer: Separate sections look good.
Small remark on the list of native connectors
- I'm not aware that we've tested s3, google_cloud_storage, azure_blob_storage, postgresql, oracle, mssql in native mode and they were working. Considering all the dependencies (drivers etc) are we confident we can support them running in native mode?
- We should not advertise connectors that are released after 8.7 or not at all: dir, sharepoint, jira. Can we take them out of the native connector list?

Reviewer 2: >I'm not aware that we've tested s3, google_cloud_storage, azure_blob_storage, postgresql, oracle, mssql in native mode and they were working. Considering all the dependencies (drivers etc) are we confident we can support them running in native mode?

No, they won't. This `config.yml` is more like the default configuration, which is supposed to be used for development. The real config used in Cloud deployment is [here](https://github.com/elastic/ent-search/blob/bfac9c6d206febc5e94ec169b1ec250d553f867a/dist/shared/cloud/coservices/python-connectors.yml#L37-L38). And you can see only `mysql` is supported for now.

Author: Indeed, but this config can be used by on-prem developers. Should we also drop them from the config file, keeping only same ones that are supported in cloud?

Reviewer: > this config can be used by on-prem developers

This is correct, in 8.7 customers with Premium license should be able to run native connectors on-prem. That was the premise of my previous comment. So the question would be 
> I'm not aware that we've tested s3, google_cloud_storage, azure_blob_storage, postgresql, oracle, mssql in native mode _on-prem_ ... are we confident we can support them running in native mode?

Reviewer: Just checked with @sphilipse and the only native connectors supported (hardcoded) in 8.7 Kibana are MongoDB and MySQL. I'd say it's worth documenting this.

Author: For now this part is annoyingly dependant on our functional tests. They are built in the way that the connectors are treated as native, so during improvements of connectors we will make it possible to remove entries from `native_connectors` list, but for now they'll have to stay.

We can remove them from the example config in this doc OR write a separate note saying that the list of actually supported native connectors might differ (we might have to update this list every release or choose a vague description).

What do you think is best?

> # Running Connectors Service in Docker

Author: I'm still super confused with it all the time 🙈 @danajuratoni It's Connector Service, right?

> # Running Connectors Service in Docker

Reviewer 3: ```suggestion
# Run Connector Service in Docker
```

> Running Connectors Service in Docker is really easy. All it requires is having Docker installed on a local system, and nothing more!

Reviewer 3: ```suggestion
To run the Connector Service in Docker, you need to have Docker installed locally.
```

This guide uses generally-available unix commands to demonstrate how to run the Connectors Service in Docker, for Windows you might have to run them in [Unix Subsystem](https://learn.microsoft.com/en-us/windows/wsl/about), rewrite the commands in PowerShell or do the steps in the guide using good old mouse and keyboard :)

> ## Steps to properly run Connector Service in Docker

Reviewer 3: ```suggestion
```

Consider deleting this subheading and using `##` headings for steps 1-4.

> 1. Create a directory that will be mounted into the Docker image.

Reviewer 3: ```suggestion
### 1. Create a directory that will be mounted into the Docker image.
```

Consider using subheading for these main steps. Same comment for steps 2-4. :)

Author: Done! I also removed the numbers 1-4 to make it look better :)

> 1. Create a directory that will be mounted into the Docker image.

Reviewer 3: ```suggestion
1. Create a directory to be mounted into the Docker image
```

> This directory will contain the configuration file that will be used to run the Connectors Service. As an example in this guide I will use home directory of the user I'm logged in as.

Reviewer 3: ```suggestion
This directory will contain the configuration file used to run the Connector Service. The examples in this guide will use the user's home directory (`~`).
```

```sh
cd ~ && mkdir connectors-python-config
```

> 2. Download a sample configuration file from this repository into newly created directory.

Reviewer 3: ```suggestion
2. Download sample configuration file from this repo.
```

> 2. Download a sample configuration file from this repository into newly created directory.

Reviewer 3: ```suggestion
## 2. Download a sample configuration file from this repository into newly created directory.
```

> Download can be done manually, or just using the line below - make sure to update `--output` argument value if your directory name is different or you want to use a different config file name.

Reviewer 3: ```suggestion
You can download the file manually, or simply run the command below. Make sure to update the `--output` argument value if your directory name is different,  or you want to use a different config file name.
```

```sh
curl https://raw.githubusercontent.com/elastic/connectors-python/main/config.yml --output ~/connectors-python-config/config.yml
```

> 3. Update the configuration downloaded to make it work for your [on-prem connector](https://www.elastic.co/guide/en/enterprise-search/current/build-connector.html#build-connector-usage)

Reviewer 3: ```suggestion
3. Update the configuration file for your [on-prem connector](https://www.elastic.co/guide/en/enterprise-search/current/build-connector.html#build-connector-usage)
```

> 3. Update the configuration downloaded to make it work for your [on-prem connector](https://www.elastic.co/guide/en/enterprise-search/current/build-connector.html#build-connector-usage)

Reviewer 3: ```suggestion
## 3. Update the configuration downloaded to make it work for your [on-prem connector](https://www.elastic.co/guide/en/enterprise-search/current/build-connector.html#build-connector-usage)
```

> In case you're running Connector Service against a dockerised version of Elasticsearch and Kibana, your config file might look the following:

Reviewer 3: ```suggestion
If you're running the Connector Service against a dockerised version of Elasticsearch and Kibana, your config file will look like this:
```

  ssl: true
  bulk:
    queue_max_size: 1024
    queue_max_mem_size: 25
    display_every: 100
    chunk_size: 1000
    max_concurrency: 5
    chunk_max_mem_size: 5
    concurrent_downloads: 10
  request_timeout: 120
  max_wait_duration: 120
  initial_backoff_duration: 1
  backoff_multiplier: 2
  log_level: info

service:
  idling: 30
  heartbeat: 300
  max_errors: 20
  max_errors_span: 600
  max_concurrent_syncs: 1
  job_cleanup_interval: 300
  log_level: INFO
  
# connector information
connector_id: <CONNECTOR_ID_FROM_KIBANA>
service_type: '<DESIRED_SERVICE_TYPE>'

sources:
  mongodb: connectors.sources.mongo:MongoDataSource
  s3: connectors.sources.s3:S3DataSource
  dir: connectors.sources.directory:DirectoryDataSource
  mysql: connectors.sources.mysql:MySqlDataSource
  network_drive: connectors.sources.network_drive:NASDataSource
  google_cloud_storage: connectors.sources.google_cloud_storage:GoogleCloudStorageDataSource
  azure_blob_storage: connectors.sources.azure_blob_storage:AzureBlobStorageDataSource
  postgresql: connectors.sources.postgresql:PostgreSQLDataSource
  oracle: connectors.sources.oracle:OracleDataSource
  mssql: connectors.sources.mssql:MSSQLDataSource
```

> Notice, that the config file you downloaded might contain more config entries, so you will need to manually copy/change the settings that apply to you. It should be sufficient to only update `elasticsearch.host` and `elasticsearch.password` to make Connectors Service run properly for you.

Reviewer 3: ```suggestion
ℹ️ The config file you downloaded might contain more lines. You will need to manually copy/change the settings that apply to you. Usually, updating `elasticsearch.host` and `elasticsearch.password` is enough for the Connector Service run properly.
```

### Running connector service in [native mode](https://www.elastic.co/guide/en/enterprise-search/current/native-connectors.html) in Docker

> In case you want to run connector service in native mode, you will need slightly different configuration file, having `native_service_types` config option populated instead of `connector_id` and `service_type`. Normally when you download a sample configuration file in step 2 you'll be able to run connector service in native mode (as long as elasticsearch host and credentials are correct).

Reviewer 3: ```suggestion
To run the connector service in native mode, you will need a slightly different configuration file. The `native_service_types` config option needs to be populated instead of `connector_id` and `service_type`. Normally, when you download a sample configuration file per step 2, you'll be able to run the Connector Service in native mode (as long as the Elasticsearch host and credentials are correct).
```

> Example of configuration file for connectors service running in native mode:

Reviewer 3: ```suggestion
Here is an example configuration file for a Connector Service running in native mode:
```

  log_level: INFO
  
# remove entries from this list to not run these connectors in this instance of the service
native_service_types:
  - mongodb
  - mysql
  - network_drive
  - s3
  - google_cloud_storage
  - azure_blob_storage
  - postgresql
  - oracle
  - dir
  - mssql

sources:
  mongodb: connectors.sources.mongo:MongoDataSource
  s3: connectors.sources.s3:S3DataSource
  dir: connectors.sources.directory:DirectoryDataSource
  mysql: connectors.sources.mysql:MySqlDataSource
  network_drive: connectors.sources.network_drive:NASDataSource
  google_cloud_storage: connectors.sources.google_cloud_storage:GoogleCloudStorageDataSource
  azure_blob_storage: connectors.sources.azure_blob_storage:AzureBlobStorageDataSource
  postgresql: connectors.sources.postgresql:PostgreSQLDataSource
  oracle: connectors.sources.oracle:OracleDataSource
  mssql: connectors.sources.mssql:MSSQLDataSource
```

After that, you can build your own docker image to run:

```
docker build -t <TAG_OF_THE_IMAGE> .
```

For example, if you've created a custom version of MongoDB connector, you can tag it with the following command:

```
docker build -t connector/custom-mongodb:1.0 .
```

> You can later use `<TAG_OF_THE_IMAGE>` instead of `docker.elastic.co/enterprise-search/elastic-connectors:8.7.0.0-SNAPSHOT` in the next step to run the docker image

Author: That's what we do now - you can actually download these images and run now, the names are already set. cc @tarekziade - what about the naming conventions that @wangch079 mentioned?

> You can later use `<TAG_OF_THE_IMAGE>` instead of `docker.elastic.co/enterprise-search/elastic-connectors:8.7.0.0-SNAPSHOT` in the next step to run the docker image

Reviewer 3: ```suggestion
You can later use `<TAG_OF_THE_IMAGE>` instead of `docker.elastic.co/enterprise-search/elastic-connectors:8.7.0.0-SNAPSHOT` in the next step to run the Docker image.
```

> 4. Run the docker image.

Reviewer 3: ```suggestion
## 4. Run the docker image.
```

> 4. Run the docker image.

Reviewer 3: ```suggestion
4. Run the Docker image.
```

Just standardising capitalisation! Might have missed a few :)

> Now you can run the docker image with Connector Service by using a command like the following:

Reviewer 3: ```suggestion
Now you can run the Docker image with the Connector Service. Here's an example command:
```

```sh
docker run \
-v ~/connectors-python-config:/config \
--network "elastic" \
--tty \
--rm \
docker.elastic.co/enterprise-search/elastic-connectors:8.7.0.0-SNAPSHOT \
/app/bin/elastic-ingest \
-c /config/config.yml
```

> You might need to adjust several things here:

Reviewer 3: ```suggestion
You might need to adjust some details here:
```

- `-v ~/connectors-python-config:/config \` - replace `~/connectors-python-config` with the directory that you've created in step 1 if you've chosen a different name for it.

> - `docker.elastic.co/enterprise-search/elastic-connectors:8.7.0.0-SNAPSHOT` - adjust the version for the connectors to match your Elasticsearch deployment version. For Elasticsearch of version 8.7 you can use connectors of the version `elastic-connectors:8.7.0.0` if you want a stable revision of the connectors, or `elastic-connectors:8.7.0.0-SNAPSHOT` if you want the latest nightly build of the connectors (not recommended). If you are using nightly builds, you will need to run `docker pull docker.elastic.co/enterprise-search/elastic-connectors:8.7.0.0-SNAPSHOT` before starting the service to make sure you're using the latest version of the docker image.

Reviewer 3: ```suggestion
- `docker.elastic.co/enterprise-search/elastic-connectors:8.7.0.0-SNAPSHOT` - adjust the version for the connectors to match your Elasticsearch deployment version. 
  - For Elasticsearch of version 8.7 you can use `elastic-connectors:8.7.0.0`for a stable revision of the connectors, or `elastic-connectors:8.7.0.0-SNAPSHOT` if you want the latest nightly build of the connectors (not recommended).
  - If you are using nightly builds, you will need to run `docker pull docker.elastic.co/enterprise-search/elastic-connectors:8.7.0.0-SNAPSHOT` before starting the service. This ensures you're using the latest version of the Docker image.
```

> - `-c /config/config.yml` - replace `config.yml` with the name of the config file you've put in the directory you've created in step 1. Normally, if you run a single instance of Connector Service you might want to leave the file name `config.yml`, but if you are planning to run multiple versions of service, then you can put configs for different service versions in single directory and update this config setting to point to different configuration files.

Reviewer 3: ```suggestion
- `-c /config/config.yml` - replace `config.yml` with the name of the config file you've put in the directory you've created in step 1. Normally, if you run a single instance of the Connector Service you can use `config.yml`. But if you(re planning to run multiple versions of the service, you can put configs for different service versions in a single directory and update this config setting to point to different configuration files.
```
